### ISSUE TITLE
dvc exp list randomly choosing tags of the same commit

### ISSUE THREAD

**Issue by tibor-mach** [iterative/dvc#9794]
This is not really a bug, but there is some inconsistency in behaviour between 

`dvc exp list` and `dvc exp show --no-pager`.

The latter seems to consistently list the experiments under the checked out branch from which they were run (even if there are unstaged changes), the former uses any of the tags associated with the commit from which these were run.

If there are more tags associated with the same commit it seems to choose the tag name at random. It is still the same commit of course, but it can be a bit confusing. 

Here's an example (using [this](https://github.com/iterative/example-get-started-experiments) repo) :

<img width="752" alt="image" src="https://github.com/iterative/dvc/assets/56956998/1167472b-489e-477a-aad0-502197757e09">

I guess it is up for discussion whether to do something about this or not (and if so which tag is the "right" one) but at least it would be good to have the discussion kept here for reference.

cc @dberenbaum

**Comment by pmrowla** [comment]
We could also make that wrapper sort tags/branches alphabetically, but that will actually affect performance for repos that have a lot of tags or branches. We have not done that sorting yet since the most common scenarios in DVC only require showing the current rev (and defaulting to the checked out branch makes us consistent in that scenario)

**Comment by dberenbaum** [comment]
I think it's also happening with `exp show`. I think the problem is that there are multiple tags referencing the same commit. See below:

```bash
$ dvc exp show
 ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────>
  Experiment           Created        train.box_loss   train.seg_loss   train.cls_loss   train.dfl_loss   metrics.precision(B)   metrics.recall(B)   metrics.mAP50(B)   metrics.m>
 ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────>
  workspace            -                           -                -                -                -                      -                   -                  -            >
  1-notebook-dvclive   Jul 29, 2023           3.2606           4.5661           3.8355           1.9272              0.0013889              0.1087         0.00077617            >
 ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────>

$ dvc exp show
 ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────>
  Experiment            Created        train.box_loss   train.seg_loss   train.cls_loss   train.dfl_loss   metrics.precision(B)   metrics.recall(B)   metrics.mAP50(B)   metrics.>
 ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────>
  workspace             -                           -                -                -                -                      -                   -                  -           >
  dvclive=best@v0.1.0   Jul 29, 2023           3.2606           4.5661           3.8355           1.9272              0.0013889              0.1087         0.00077617           >
 ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────>
```

It also impacts the VS Code extension (cc @mattseddon):


https://github.com/iterative/dvc/assets/2308172/eab681fe-08a6-4739-9bec-387ed3fe7e04


https://github.com/iterative/dvc/assets/2308172/41d8224e-3a46-4d4f-bbb9-384ebd1dcfbe

**Comment by mattseddon** [comment]
Seems like https://github.com/iterative/dvc/issues/6051

**Comment by dberenbaum** [comment]
Yep, looks like a continuation of #6051, although I'm fine to keep discussion going here. 

I agree with this:

> Ideally all aliases (refs with the same `HEAD`) should be shown, or at least the same one consistently (e.g. first alphabetically. I don't think Git keepa a record of when an alias branch was created so no way to use the last one created, I think).

_Originally posted by @casperdcl in https://github.com/iterative/dvc/issues/6051#issuecomment-847029454_

Is there a blocker to doing this?

**Comment by pmrowla** [comment]
There is nothing preventing us from doing that but it will be slower than what we are doing now (it's just what I posetd in the earlier comment https://github.com/iterative/dvc/issues/9794#issuecomment-1663949030)

**Comment by dberenbaum** [comment]
Okay, got confused and thought that comment was about sorting, but I guess it's really about retrieving lots of refs.

**Comment by dberenbaum** [comment]
Let me keep this issue restricted to making `dvc exp list` match `dvc exp show` and reopen #6051 for the larger issue (both commands randomly choose tags for non-HEAD revisions or in detached HEAD state).

**Comment by dberenbaum** [comment]
@tibor-mach No pressure, but do you have interest and time in taking a shot at what @pmrowla suggested?

**Comment by tibor-mach** [comment]
It's been a rainy weekend round here, so I gave it a go.

It turns out that `dvc exp list` needs a different string format than `dvc exp show`. I did not want to duplicate the wrapper so I added an option to switch between the two but this is perhaps what made if fail benchmarks (I use `re.sub` ). Other than that it seems to work as intended.

It still needs some code cleaning though, it would be great if @pmrowla could help me with that. I made some comments in the code in the PR as well.
